### DISCUSSION TITLE
Multiple spindle control [PWM; M67,M68]

### DISCUSSION THREAD

**Discussion by martinjurekcz** [grblHAL/core#142] [Ideas]
Hi everyone,
Im looking for a control software for my new laser engraver. I designed working head with 2 laser modules (view of the cut attached) (6W each) for raster engraving photos on wood. Since the modules are tiny (8mm) they can be stacked side by side in one working head and produce multiple rows within one pass. (2 lasers = 2x faster production, 3lasers...). Is grblHAL up to this task? I noticed that there should be support for M67, M68 gcodes (analog output - PWM), however I don't know how to physically implement it. Is there a driver board with sufficient numbers of pins and interrupts that could drive additional PWM as a second or third spindle via M67, M68 codes?
thanks for any tips and information on how to achieve that.
![jd_laser_hlava](https://user-images.githubusercontent.com/60178361/168604894-22ddfaab-1a89-4b6e-add9-d4b7ae3dc59f.png)

**Comment by terjeio** [comment]
> Is grblHAL up to this task?

Yes and no... It will be possible to add a plugin (or plugins) to achieve this. I would not use M67 or M68 but rather go for a binary protocol instead (for speed). There is an ongoing discussion for that [here](https://github.com/arkypita/LaserGRBL/discussions/1818).

---

I have started work on multi spindle support in the core but do not know if/when this will be released. This _could_ form part of a solution.
An example for how to add anlogue outputs can be found [here](https://github.com/grblHAL/iMXRT1062/blob/master/grblHAL_Teensy4/src/ioports.c), it is for an off-chip MCP3221 DAC. SiImilar code can be used for PWM or on-chip DACs and makes the outputs available from plugin code and/or M67,M68.
